lib/client.js

```diff
@@ -114,62 +154,102 @@ module.exports = function (dependencies) {
 
-        if (status === 200) {
-          resolve({ device });
-        } else if (responseData !== "") {
-          const response = JSON.parse(responseData);
+          if (status === 200) {
+            resolve({ device });
+          } else if (responseData !== "") {
+            const response = JSON.parse(responseData);
 
-          if (status === 403 && response.reason === "ExpiredProviderToken" && retryCount < 2) {
-            this.config.token.regenerate(tokenGeneration);
-            resolve(this.write(notification, device, retryCount + 1));
-            return;
-          } else if (status === 500 && response.reason === "InternalServerError") {
-            this.session.destroy();
-            let error = new VError("Error 500, stream ended unexpectedly");
+            if (status === 403 && response.reason === "ExpiredProviderToken" && retryCount < 2) {
+              this.config.token.regenerate(tokenGeneration);
+              resolve(this.write(notification, device, retryCount + 1));
+              return;
+            } else if (status === 500 && response.reason === "InternalServerError") {
+              this.closeAndDestroySession();
+              let error = new VError("Error 500, stream ended unexpectedly");
+              resolve({ device, error });
+              return;
+            }
+
+            resolve({ device, status, response });
+          } else {
+            let error = new VError("stream ended unexpectedly");
             resolve({ device, error });
-            return;
           }
-
-          resolve({ device, status, response });
-        } else {
-          let error = new VError("stream ended unexpectedly");
+        } catch (e) {
+          const error = new VError(e, 'Unexpected error processing APNs response');
+          if (logger.enabled) {
+            logger(`Unexpected error processing APNs response: ${e.message}`);
+          }
           resolve({ device, error });
         }
-      })
+      });
+
+      request.setTimeout(this.config.requestTimeout, () => {
+        if (logger.enabled) {
+          logger('Request timeout');
+        }
+
+        request.close(NGHTTP2_CANCEL);
+
+        resolve({ device, error: new VError("apn write timeout") });
+      });
+
+      request.on("aborted", () => {
+        if (logger.enabled) {
+          logger('Request aborted');
+        }
+
+        resolve({ device, error: new VError("apn write aborted") });
+      });
 
       request.on("error", (error) => {
         if (logger.enabled) {
           logger(`Request error: ${error}`);
         }
 
         if (typeof error === "string") {
-          error = new VError("apn write failed: %s", err);
+          error = new VError("apn write failed: %s", error);
         } else {
           error = new VError(error, "apn write failed");
         }
         resolve({ device, error });
       });
 
+      if (logger.enabled) {
+        request.on("close", () => {
+          logger("Request closed");
```

Comment:
nit: This seems harmless but excessively verbose - It seems at a glance like any request would emit both 'end' and 'close' whether it succeeded or failed. So we'd already be indicating this with logs such as ` Request ended with status null and responseData`

e.g. running the unit tests with DEBUG=apn

```
  apn Session connected +0ms
  apn Request ended with status null and responseData:  +12ms
  apn Request closed +0ms
  apn Request ended with status null and responseData:  +0ms
  apn Request closed +0ms
  apn Request ended with status null and responseData:  +0ms
  apn Request closed +0ms
  apn GOAWAY received: (errorCode 0, lastStreamId: 7, opaqueData: undefined) +0ms
  apn Request ended with status null and responseData:  +1ms
  apn Request closed +0ms
  apn Session closed +0ms

```

Reply:
ok. agree. I've just removed.